src/dune_engine/dep.mli

```diff
@@ -110,19 +110,28 @@ module Set : sig
   include
     Set.S with type elt = t and type 'a map := 'a Map.t and type t = unit Map.t
 
-  (** Return dependencies on all source files under a certain source directory.
-
-      Dependency on a source_tree requires special care for empty directory, so
-      you should use this function rather than manually traverse the source
-      tree. *)
-  val source_tree : Path.t -> t Memo.t
-
-  (** Same as [source_tree] but also return the set of files as a set. Because
-      of the special care for empty directories, the set of dependencies
-      returned contains dependencies other than [File]. So extracting the set of
-      files from the dependency set is a bit awkward. This is why this function
-      exist. *)
-  val source_tree_with_file_set : Path.t -> (t * Path.Set.t) Memo.t
+  module Tree
+      (Monad : Monad.S) (_ : functor (Outcome : Monoid.S) -> sig
+        val map_reduce :
+             Path.t
+          -> f:(path:Path.t -> files:String.Set.t -> Outcome.t Monad.t)
+          -> Outcome.t Monad.t
+      end) : sig
+    (** Return dependencies on all source files under a certain source
+        directory.
+
+        Dependency on a source_tree requires special care for empty directory,
```

Comment:
I don't understand what this special care is. It seems worth clarifying. (I know it's an old comment but we might as well refresh it a bit while changing this code.)

Reply:
(Maybe the empty directory case is treated specially as an optimisation but then I'm not sure it's important to mention this at this level of the API?)

Reply:
The comment in the .ml file clarifies it:

```
  (* This is to force the rules to be loaded for directories without files when
     depending on [(source_tree x)]. Otherwise, we wouldn't clean up stale
     directories in directories that contain no file. *)
```

Should the reader be told to refer to the .ml? Or move this comment to the .mli?

Reply:
Oh, I see. I think it makes sense to keep this comment at the current location in the `ml` file (though please don't forget to rename `source_tree` to `Source_tree.files`), but it does also seem useful to say in the `mli` that empty directories are depended on appropriately to ensure stale directories are cleaned from garbage.